src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py

```diff
@@ -0,0 +1,1023 @@
+# synthcity relative
+from .modules import MLPDiffusion, ResNetDiffusion
+from .utils import (
+    FoundNANsError,
+    discretized_gaussian_log_likelihood,
+    extract,
+    index_to_log_onehot,
+    log_1_min_a,
+    log_add_exp,
+    log_categorical,
+    mean_flat,
+    normal_kl,
+    ohe_to_categories,
+    sliced_logsumexp,
+    sum_except_batch,
+)
+
+
+def get_named_beta_schedule(
+    schedule_name: str, num_diffusion_timesteps: int
+) -> np.ndarray:
+    """
+    Get a pre-defined beta schedule for the given name.
+    The beta schedule library consists of beta schedules which remain similar
+    in the limit of num_diffusion_timesteps.
+    Beta schedules may be added, but should not be removed or changed once
+    they are committed to maintain backwards compatibility.
+    """
+    if schedule_name == "linear":
+        # Linear schedule from Ho et al, extended to work for any number of
+        # diffusion steps.
+        scale = 1000 / num_diffusion_timesteps
+        beta_start = scale * 0.0001
+        beta_end = scale * 0.02
+        return np.linspace(
+            beta_start, beta_end, num_diffusion_timesteps, dtype=np.float64
+        )
+    elif schedule_name == "cosine":
+        return betas_for_alpha_bar(
+            num_diffusion_timesteps,
+            lambda t: math.cos((t + 0.008) / 1.008 * math.pi / 2) ** 2,
+        )
+    else:
+        raise NotImplementedError(f"unknown beta schedule: {schedule_name}")
+
+
+def betas_for_alpha_bar(
+    num_diffusion_timesteps: int, alpha_bar: Callable, max_beta: float = 0.999
+) -> np.ndarray:
+    """
+    Create a beta schedule that discretizes the given alpha_t_bar function,
+    which defines the cumulative product of (1-beta) over time from t = [0,1].
+    :param num_diffusion_timesteps: the number of betas to produce.
+    :param alpha_bar: a lambda that takes an argument t from 0 to 1 and
+                      produces the cumulative product of (1-beta) up to that
+                      part of the diffusion process.
+    :param max_beta: the maximum beta to use; use values lower than 1 to
+                     prevent singularities.
+    """
+    betas = []
+    for i in range(num_diffusion_timesteps):
+        t1 = i / num_diffusion_timesteps
+        t2 = (i + 1) / num_diffusion_timesteps
+        betas.append(min(1 - alpha_bar(t2) / alpha_bar(t1), max_beta))
+    return np.array(betas)
+
+
+class GaussianMultinomialDiffusion(torch.nn.Module):
+    def __init__(
+        self,
+        num_numerical_features: int,
+        num_categorical_features: tuple,
+        model_type: str = "mlp",
+        model_params: Optional[dict] = None,
+        num_timesteps: int = 1000,
+        gaussian_loss_type: str = "mse",
+        gaussian_parametrization: str = "eps",
+        multinomial_loss_type: str = "vb_stochastic",
+        parametrization: str = "x0",
+        scheduler: str = "cosine",
+        device: torch.device = torch.device("cpu"),
+        verbose: int = 0,
```

Comment:
Instead of the `verbose` arg, please use the `synthcity.logger` module, with `log.error`/`log.debug` etc. You can find examples in the code